vllm/v1/worker/gpu_model_runner.py

```diff
@@ -701,34 +705,45 @@ def _calc_mrope_positions(self, scheduler_output: "SchedulerOutput"):
             assert num_scheduled_tokens == prompt_part_len + completion_part_len
 
             if prompt_part_len > 0:
-                # prompt's mrope_positions are pre-computed
+                assert req.mrope_positions is not None
                 dst_start = mrope_pos_ptr
-                dst_end = mrope_pos_ptr + prompt_part_len
+                dst_end = dst_start + prompt_part_len
                 src_start = num_computed_tokens
-                src_end = num_computed_tokens + prompt_part_len
-
-                self.mrope_positions_cpu[:, dst_start:dst_end] = \
-                    req.mrope_positions[:,src_start:src_end]
-
+                src_end = src_start + prompt_part_len
+                self.mrope_positions_cpu[:, dst_start:dst_end] =\
+                    req.mrope_positions[:, src_start:src_end]
                 mrope_pos_ptr += prompt_part_len
 
             if completion_part_len > 0:
-                # compute completion's mrope_positions on-the-fly
                 dst_start = mrope_pos_ptr
-                dst_end = mrope_pos_ptr + completion_part_len
-
-                self.mrope_positions_cpu[:, dst_start:dst_end] = \
-                    MRotaryEmbedding.get_next_input_positions_tensor(
-                        req.mrope_position_delta,
-                        context_len=num_computed_tokens +
-                        prompt_part_len,
-                        seq_len=num_computed_tokens +
-                        prompt_part_len +
-                        completion_part_len,
-                    )
-
+                context_len = num_computed_tokens + prompt_part_len
+                seq_len = context_len + completion_part_len
+                comp_dst_slices.append(
+                    (dst_start, dst_start + completion_part_len))
+                comp_data.append((req.mrope_position_delta + context_len,
+                                  req.mrope_position_delta + seq_len))
                 mrope_pos_ptr += completion_part_len
 
+        # NOTE (Kevin C): mmprovement came from premaking the calculated mrope
+        # positions, the copying over the slices
+
+        # Process computed positions
+        if comp_data:
+            mrope_min = min(data[0] for data in comp_data)
+            mrope_max = max(data[1] for data in comp_data)
+
+            # make sure that this premade tensor covers all possible range
+            base_arange = MRotaryEmbedding.get_next_input_positions_tensor(
+                0,
+                context_len=mrope_min,
+                seq_len=mrope_max + 1,
+            )
+            for dst_slice, data in zip(comp_dst_slices, comp_data):
+                dst_start, dst_end = dst_slice
+                if data[0] != data[1]:
+                    self.mrope_positions_cpu[:, dst_start:dst_end] = \
+                        base_arange[:, data[0] - mrope_min: data[1] - mrope_min]
```

Comment:
```suggestion
                if data[0] != data[1]:
                    src_start, src_end = data[0] - mrope_min, data[1] - mrope_min
                    self.mrope_positions_cpu[:, dst_start:dst_end] = \
                        base_arange[:, src_start: src_end]
```
I think it would be more readable if we can move calculation out of the slice expression